client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,262 @@
+//! Registry>`. That can be used as a `futures::Stream` to receive the messages.
+//! Upon drop the `Receiver<Message, Registry>` shall unregister itself from the `Hub`.
+
+use std::{
+	collections::HashMap,
+	ops::DerefMut,
+	pin::Pin,
+	sync::{Arc, Weak},
+	task::{Context, Poll},
+};
+
+use ::futures::stream::{FusedStream, Stream};
+use ::parking_lot::Mutex;
+
+use crate::mpsc::{TracingUnboundedReceiver, TracingUnboundedSender};
+
+// pub mod channels;
+
+/// The type to identify subscribers.
+pub type SubsID = crate::id_sequence::SeqID;
+
+/// Unsubscribe: unregisters a previously created subscription.
+pub trait Unsubscribe {
+	/// Remove all registrations of the subscriber with ID `subs_id`.
+	fn unsubscribe(&mut self, subs_id: &SubsID);
+}
+
+/// Subscribe using a key of type `K`
+pub trait Subscribe<K> {
+	/// Register subscriber with the ID `subs_id` as having interest to the key `K`.
+	fn subscribe(&mut self, subs_key: K, subs_id: SubsID);
+}
+
+/// Dispatch a message of type `M`.
+pub trait Dispatch<M> {
+	/// The type of the that shall be sent through the channel as a result of such dispatch.
+	type Item;
+	/// The type returned by the `dispatch`-method.
+	type Ret;
+
+	/// Dispatch the message of type `M`.
+	///
+	/// The implementation is given an instance of `M` and is supposed to invoke `dispatch` for
+	/// each matching subscriber, with an argument of type `Self::Item` matching that subscriber.
+	///
+	/// Note that this does not have to be of the same type with the item that will be sent through
+	/// to the subscribers. The subscribers will receive a message of type `Self::Item`.
+	fn dispatch<F>(&mut self, message: M, dispatch: F) -> Self::Ret
+	where
+		F: FnMut(&SubsID, Self::Item);
+}
+
+/// A subscription hub.
+///
+/// Does the subscription and dispatch.
+/// The exact subscription and routing behaviour is to be implemented by the Registry (of type `R`).
+/// The Hub under the hood uses the channel defined in `crate::mpsc` module.
+#[derive(Debug)]
+pub struct Hub<M, R> {
+	tracing_key: &'static str,
+	shared: Arc<Mutex<Shared<R, TracingUnboundedSender<M>>>>,
+}
+
+/// The receiving side of the subscription.
+///
+/// The messages are delivered as items of a `futures::Stream`.
+/// Upon drop this receiver unsubscribes itself from the `Hub`.
+#[derive(Debug)]
+pub struct Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	// NB: this field should be defined before `rx`.
+	// (The fields of a struct are dropped in declaration order.)[https://doc.rust-lang.org/reference/destructors.html]
+	_unsubs_guard: UnsubscribeGuard<R, TracingUnboundedSender<M>>,
+
+	// NB: this field should be defined after `_unsubs_guard`.
+	rx: TracingUnboundedReceiver<M>,
+}
+
+#[derive(Debug)]
+struct UnsubscribeGuard<R, Tx>
+where
+	R: Unsubscribe,
+{
+	shared: Weak<Mutex<Shared<R, Tx>>>,
+	subs_id: SubsID,
+}
+
+#[derive(Debug)]
+struct Shared<R, Tx> {
+	id_sequence: crate::id_sequence::IDSequence,
+	registry: R,
+	sinks: HashMap<SubsID, Tx>,
```

Comment:
So this essentially needs to track/manage all of the subscribers twice, right? Once here in this `HashMap`, and once again by whoever's using this. Isn't that somewhat of a half measure? (:

If we don't want to go for what was originally proposed (that is - only adding a channel which notifies the sender out-of-band that the receiver was dropped so that it can remove the subscription) and want to create a more encompassing abstraction which would manage the subscriptions then I think it'd make sense to actually have it fully manage those subscriptions.

Reply:
Handling the channels externally (in relation to the Registry implementation) indeed introduces another indirection when choosing a channel.

In my opinion that is a reasonable price to keep the Registry implementation unaware of the nature of the channels (or for that matter parametrising the Registry with a type of a channel).